hfs/helpers.py

```diff
@@ -155,16 +122,50 @@ def shrink_dag(x_identifier, digraph):
     ----------
     digraph : networkx.DiGraph
             The resulting DAG after removing all irrelevant leaf nodes.
-
     """
-    leaves = get_irrelevant_leaves(x_identifier, digraph)
-    while leaves:
-        for x in leaves:
-            digraph.remove_node(x)
-        leaves = get_irrelevant_leaves(x_identifier, digraph)
+    to_remove = {
+        node
+        for node in digraph.nodes()
+        if _is_irrelevant_leaf(node, x_identifier, digraph)
+    }
+
+    while to_remove:
+        # Recompute the list of nodes to check (predecessors of removed nodes)
+        to_check = {pred for node in to_remove for pred in digraph.predecessors(node)}
+        digraph.remove_nodes_from(to_remove)
+        to_remove = {
+            node for node in to_check if _is_irrelevant_leaf(node, x_identifier, digraph)
+        }
+
     return digraph
 
 
+def _is_irrelevant_leaf(node, x_identifier, digraph):
```

Comment:
Just a comment to ourselves that we need to add in the documentation (and potentially in a future CONTRIBUTING.md) that we do not use type hints in private functions – at least for now.

Reply:
Do you mean that we should remove the parameters from the docstring here?

Reply:
Solved.